sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullFunctions.scala

```diff
@@ -68,22 +81,35 @@ case class Coalesce(children: Seq[Expression]) extends Expression {
     """ +
     children.map { e =>
       val eval = e.gen(ctx)
-      s"""
-        if (${ev.isNull}) {
-          ${eval.code}
-          if (!${eval.isNull}) {
-            ${ev.isNull} = false;
-            ${ev.primitive} = ${eval.primitive};
-          }
-        }
-      """
+      dataType match {
+        case DoubleType | FloatType =>
+          s"""
+            if (${ev.isNull}) {
+              ${eval.code}
+              if (!${eval.isNull} && !Double.isNaN(${eval.primitive})) {
```

Comment:
I think this will introduce a conversion from Float to Double (they have different binary format), so I'd like to avoid them (using different branch for them).